docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+
+- Maintains an in-memory collection of upkeepResults. It should maintain a single result per `upkeepTriggerID`. 
+    - Overwrites results for duplicated `upkeepTriggerID`.
+- Each result has a TTL. When the TTL expires
+    - conditional upkeeps will be sampled and checked again automatically
+    - log upkeeps that were missed are expected to be picked up by the log recoverer **TBD**
+- Provides a read API to view results, which takes as input (`pseudoRandomSeed`, `limit`). Sorts all keys (trigger, upkeepID) with the `seed` and provides first `limit` results. We do not do FIFO here as potentially out of sync old results will block new results sent by the node for consensus.
+- Provides an API to remove (trigger, upkeepID) from the store
+
+### Metadata Store
+
+This component stores the metadata to be sent within observations to the network. There are three categories of metadata
+
+- Conditional Upkeep Instructions
+- Log recovery Instructions
+- Latest block history
+
+Every item has a TTL, and upon expiry items are just removed without any action.
+
+The store provides an add / view / remove API for other components.
+
+### Samples Observer
+
+Processes samples of upkeeps to be checked, using the latest block number as a trigger. It does the following procedures:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **metadata store**
+with conditional Upkeep instructions
+- Errors and ineligible results are ignored
+
+### Conditional Observer
+
+Processes coordinated block + upkeep payload coming from plugin and does the following:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **result store**
+- Errors and ineligible results are ignored
+
+### Log Observer
+
+Processes latest logs for active upkeeps. It does the following procedures:
+
+- Called every second by Log Trigger Ticker with latest logs as full payloads.
+- Filter out logs with block number that is older than threshold
+    - These are expected to be picked up by the log recoverer 
+- Pre-processes to filter the logs already present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into result store
+- If runner gave an error, put back into log retry ticker, to be performed after a fixed timeout (can extend to exponential backoff)
+
+<aside>
+💡 Note: Duplicate logs can enter this flow, this component should be able to handle that. This component can maintain in-memory state for processed logs and can also use that state for exponential backoff
+</aside>
+
+### Retry Observer
+
+**TODO**
```

Comment:
The retry observer uses a schedule ticker as the driver with a coordinator preprocessor and runner. For each payload that enters the retry observer a wait time of 10 seconds is applied before being forwarded on to the runner. In postprocessing, if a result is retryable and the max retry time of 5 minutes has not passed, it is added back to the same retry observer to wait again. If the max retry time has elapsed, the result is passed to a recovery proposal observer. If the result is eligible, it is sent to the result store.